<?xml version="1.0"?>
<launch>
	<!-- Launch file for PX4 and IBVS -->
	<!-- JSW May 2018 -->

    <!--  -->
    <!-- ARGS -->
    <!--  -->
    <arg name="test_name" />

    <!-- Target Location args -->
    <arg name="tn" />
    <arg name="te" />

    <!-- State Machine args -->
    <arg name="u_max_outer" value="0.5" />
    <arg name="v_max_outer" value="0.5" />
    <arg name="w_max_outer" value="0.7" />

    <arg name="u_max_inner" value="0.2" />
    <arg name="v_max_inner" value="0.2" />
    <arg name="w_max_inner" value="0.2" />

    <arg name="rendezvous_height" value="13.1" />
    <arg name="wp_threshold" value="1.0" />

    <arg name="landing_distance_threshold" value="0.5" />
    <arg name="p_des_error_outer_threshold" value="50.0" />

    <!-- IBVS Control Gains -->
    <arg name="lambda_vx" value="0.5" />
    <arg name="lambda_vy" value="0.5" />
    <arg name="lambda_vz" value="0.7" />
    <arg name="lambda_wz" value="0.5" />

    <!-- rosbag args -->
    <arg name="record_rosbag" default="false" />

    <!-- save_mat_data args -->
    <arg name="save_mat_data" default="false" />

    <!-- Point Grey Camera Args -->
    <arg name="frame_rate" value="30" />
    <arg name="calibrated" default="true" />

    <!-- ArUco args -->
    <arg name="show_aruco_frame" value="false" />
	
    
    <!--  -->
    <!-- PX4 -->
    <!--  -->

    <!-- Launch PX4 pixhawk (taken from mavros px4.launch) -->
	<arg name="fcu_url" default="/dev/ttyTHS2:921600" />
	<arg name="gcs_url" default="" />
	<arg name="tgt_system" default="1" />
	<arg name="tgt_component" default="1" />
	<arg name="log_output" default="screen" />

	<include file="$(find mavros)/launch/node.launch">
	    <arg name="pluginlists_yaml" value="$(find mavros)/launch/px4_pluginlists.yaml" />
		<arg name="config_yaml" value="$(find mavros)/launch/px4_config.yaml" />

		<arg name="fcu_url" value="$(arg fcu_url)" />
		<arg name="gcs_url" value="$(arg gcs_url)" />
		<arg name="tgt_system" value="$(arg tgt_system)" />
		<arg name="tgt_component" value="$(arg tgt_component)" />
		<arg name="log_output" value="$(arg log_output)" />
	</include>


	<!--  -->
    <!-- MAVROS NED -->
    <!--  -->

    <!-- Node to convert mavros data to NED estimate data -->
	<node name="mavros_ned" pkg="ibvs_sim" type="mavros_ned" output="screen" />


    <!--  -->
    <!-- IBVS Control -->
    <!--  -->

    <rosparam command="load" file="$(find ibvs_sim)/params/ibvs_positioning_high.yaml"/>

    <!-- p_des Publisher -->
    <node name="p_des_mapper" pkg="ibvs_sim" type="p_des_mapper.py" output="screen">
        <remap from="/quadcopter/camera/camera_info" to="/camera_info" />
    </node>
    
    <!-- IBVS Outer -->
    <node name="ibvs" pkg="ibvs_sim" type="ibvs_adaptive.py" output="screen">

        <param name="lambda_vx" value="$(arg lambda_vx)"/>
        <param name="lambda_vy" value="$(arg lambda_vy)"/>
        <param name="lambda_vz" value="$(arg lambda_vz)"/>
        <param name="lambda_wz" value="$(arg lambda_wz)"/>

        <remap from="/ibvs/pdes" to="/ibvs/pdes_outer"/>
        <remap from="/aruco/marker_corners" to="/aruco/marker_corners_outer"/>
        <remap from="/aruco/distance" to="/aruco/distance_outer"/>
        <remap from="/ibvs/ibvs_error" to="/ibvs/ibvs_error_outer" />

        <remap from="/quadcopter/estimate" to="/mavros_ned/estimate" />
        <remap from="/quadcopter/camera/camera_info" to="/camera_info" />
    </node>


    <!--  -->
    <!-- IBVS State Machine -->
    <!--  -->

    <node name="ibvs_state_machine" pkg="ibvs_sim" type="ibvs_state_machine_uw.py" output="screen" >

        <param name ="mode" value="mavros" />

    	<param name="u_max" value="$(arg u_max_outer)" />
    	<param name="v_max" value="$(arg v_max_outer)" />
    	<param name="w_max" value="$(arg w_max_outer)" />

        <param name="u_max_inner" value="$(arg u_max_inner)" />
        <param name="v_max_inner" value="$(arg v_max_inner)" />
        <param name="w_max_inner" value="$(arg w_max_inner)" />

        <param name="rendezvous_height" value="$(arg rendezvous_height)" />
        <param name="wp_threshold" value="$(arg wp_threshold)" />

        <param name="landing_distance_threshold" value="$(arg landing_distance_threshold)" />
        <param name="p_des_error_outer_threshold" value="$(arg p_des_error_outer_threshold)" />

        <param name="test_name" value="$(arg test_name)" />

        <remap from="estimate" to="/mavros_ned/estimate" />
        <remap from="/quadcopter/camera/camera_info" to="/camera_info" />
    </node>


    <!--  -->
    <!-- Target EKF -->
    <!--  -->

    <node name="target_ekf" pkg="ibvs_sim" type="target_ekf.py" output="screen">

        <param name="delta_z" value="0.08" />

        <remap from="/quadcopter/euler" to="/mavros_ned/euler" />
        <remap from="/quadcopter/ground_truth/odometry/NED" to="/mavros_ned/estimate" />

    </node>

    
    <!--  -->
    <!-- ArUco Localization -->
    <!--  -->

    <node pkg="aruco_localization" type="aruco_localization" name="aruco" output="screen">
        <param name="show_output_video" value="$(arg show_aruco_frame)" />
        <param name="markermap_config" value="$(find ibvs_sim)/params/nested148_65.yaml" />
        <!-- <param name="marker_size" value="0.1760" /> -->
        <param name="marker_size" value="0.7071" />
        <param name="marker_size_inner" value="0.125" />
        <param name="id_outer" value="148" />
        <param name="id_inner" value="65" />

        <param name="draw_data" value="false" />

        <param name="shrink_image" value="true" />

        <param name="camera_offset_x" value="0.0" />
        <param name="camera_offset_y" value="0.0" />
        <param name="camera_offset_z" value="0.08" />

        <param name="debug_save_input_frames" value="false" />
        <param name="debug_save_output_frames" value="false" />
        <param name="debug_image_path" value="/tmp/arucoimages" />

        <remap from="input_image" to="/image_raw" />
        <remap from="output_image" to="aruco/image" />
        <remap from="/quadcopter/ground_truth/odometry/NED" to="/mavros_ned/estimate" />
    </node>


    <!--  -->
    <!-- Point Grey Camera -->
    <!--  -->

    <node name="pointgrey_camera" pkg="pointgrey_camera_driver" type="camera_node">

        <param name="camera_info_url" if="$(arg calibrated)"
            value="file://$(find aruco_localization)/camera_info/llnl_chameleon_resized_962x720.yaml" />
        <param name="frame_rate" value="$(arg frame_rate)" />
        <param name="auto_shutter" value="true" />
        <!-- <param name="shutter_speed" value="0.03" /> -->
        <param name="format7_color_coding" value="mono8" />
        <param name="auto_white_balance" value="true" />
    </node>


    <!--  -->
    <!-- Target Publisher -->
    <!--  -->

    <node name="target_publisher" pkg="ibvs_sim" type="target_pub.py" output="screen">
        <param name="target_N" value="$(arg tn)" />
        <param name="target_E" value="$(arg te)" />
    </node>


    <!--  -->
    <!-- Mat Saver -->
    <!-- -->

    <group if="$(arg save_mat_data)">
        <node name="data_saver" pkg="ibvs_sim" type="save_mat_data.py" output="screen">
            <param name="filename" value="$(arg test_name)" />
            <remap from="/quadcopter/estimate" to="/mavros_ned/estimate" />
        </node>
    </group>

    
    <!--  -->
    <!-- rosbag record -->
    <!--  -->

    <group if="$(arg record_rosbag)">
        <node name="record" pkg="rosbag" type="record" args="/mavros/local_position/pose /mavros/local_position/odom /mavros/local_position/velocity /mavros_ned/estimate /mavros_ned/euler /aruco/marker_corners_outer /aruco/marker_corners_inner /camera_info /aruco/image/compressed /aruco/estimate /aruco/distance_inner /aruco/distance_outer /aruco/heading_outer /quadcopter/high_level_command /mavros/setpoint_raw/local /quadcopter/attitude_avg /quadcopter/ibvs_active /status_flag /ibvs_status_flag /ibvs/vel_cmd /ibvs_inner/vel_cmd /aruco/orientation_inner /ibvs/ibvs_error_outer /ibvs/ibvs_error_inner /target_position /target_ekf/velocity /target_ekf/velocity_lpf -o $(arg test_name)" />
    </group>



</launch>
